### PR TITLE
Refactor and update

### PR DESCRIPTION
**Pull request by MollsReis**
Please feel free to ask tons of questions and add comments. I will be back next week to answer any questions and address any concerns before a potential merge.

### CODE REVIEW

**Comment by MollsReis** [comment]
Looks like we'll need to update for travis.yml and bring in [VCR](https://github.com/vcr/vcr) to mock web calls.

**Comment by sckott** [comment]
@EvilScott thanks for this

> Looks like we'll need to update for travis.yml and bring in VCR to mock web calls.

yeah

**Comment by sckott** [comment]
We can ignore travis for this PR, unless you want to add some tests

**Comment by MollsReis** [comment]
I added some tests and brought back `travis.yml`, give me a little bit and I think I can push another commit to fix stuff

**Comment by sckott** [comment]
:+1:

File: api.rb [hunk 1]
```diff
@@ -1,684 +1,139 @@
...
+  before do
+    # set headers
+    headers 'Content-Type' => 'application/json; charset=utf8'
+    headers 'Access-Control-Allow-Methods' => 'HEAD, GET'
+    headers 'Access-Control-Allow-Origin' => '*'
+    cache_control :public, :must_revalidate, max_age: 60
+
+    # use redis caching
+    if $config['caching']
+      @cache_key = Digest::MD5.hexdigest(request.url)
```
Comment (sckott) [inline api.rb:30]: Is the `request.url` just the base url (e.g., `fishbase.ropensci.org`)?

**Comment by sckott** [comment]
Does this PR include protection against SQL injection?  I wasn't sure what is best practice to prevent it - I imagine you've considered it in your code

see #7

**Comment by MollsReis** [comment]
Yeah using an ORM protects against injection well by the nature of the code

**Comment by MollsReis** [comment]
So getting the tests to run independent of a DB connection is a little more complicated than I initially thought. I'll set up some test fixtures when I get back in next week.

File: api.rb (hunk 1 above)
Comment (MollsReis) [inline api.rb:30]: Its actually the full URL, ex: `http://fishbase.ropensci.org/taxa/?limit=10`

Docs here: http://www.sinatrarb.com/intro.html#Accessing%20the%20Request%20Object

**Comment by sckott** [comment]
Sounds good

File: api.rb (hunk 1 above)
Comment (sckott) [inline api.rb:30]: Sweet. Curious if you think we'd also want to take into account headers, or at least some headers - no reason I can think of right now, given I don't think we require an headers passed

**Comment by MollsReis** [comment]
ok tests are updated. they could use more work in the future but that will require us to make some factories to generate fixtures to test against ([factory_girl](https://github.com/thoughtbot/factory_girl) is my preferred tool for this).

If you or your users want to try my version at https://github.com/FiN-FBSLB/fishbaseapi/ for a few days and let me know any issues or questions they encounter we can address any other "gotchas" before they cause your users any headaches post-merge

**Comment by sckott** [comment]
Do you have your version running on a public address anywhere?

**Comment by sckott** [comment]
@EvilScott how do you test this locally?  Do you run in docker?

**Comment by MollsReis** [comment]
@sckott  sorry for the slow response. to answer your questions:
1. nope, we don't have this hosted publicly yet (that will probably be a while)
2. to run the tests just `rake` from the root of the project, i stubbed out the DB dependency

**Comment by sckott** [comment]
1. Okay
2. Thanks, ill try that.

**Comment by sckott** [comment]
@EvilScott  woops, sorry, I meant how do you run your version locally? Before I would get mysql and redis up, then run `thin start`, or do via unicorn

**Comment by sckott** [comment]
nevermind, got it locally

**Comment by sckott** [comment]
Nice work on this :) Much simpler for sure. 

After running locally, a few questions/thoughts. Some of these changes (if you agree with them) we could deal with after merge, or change before if you are so inclined:
- 404 err: other errors are given back in JSON.  I'd kinda prefer all errors to be given back in JSON, not plain text, i guess its just a matter of changing https://github.com/FiN-FBSLB/fishbaseapi/blob/master/api.rb#L44-L46 - Is there a good reason to give back 404 as plain text?
- minor thing: lowercase routes in output of `/heartbeat` as Upcased routes won't work e.g,. `/Species`
- AFAICT the `fields` param (to only give back fields required by user) we used to have is ignored, or not included. Am I missing something?
- Error messages used to be more specific/useful. E.g. request to `http://127.0.0.1:8080/genera/asd` now gives `"FBApp::Error: No results found"`, whereas it used to give `id must be an integer`
- methods not allowed  (405 errors) was dropped, https://github.com/ropensci/fishbaseapi/blob/master/api.rb#L439-L442

**Comment by MollsReis** [comment]
1. We can change easily, though typically error status is not in JSON as a practice.
2. I'll fix this
3. Can you give an example?
4. Is there a list of specific error messages that are useful that we should put in?
5. We can add these back in

**Comment by sckott** [comment]
> Can you give an example?

http --verbose 'http://127.0.0.1:8080/species/?genus=Aborichthys&limit=2&fields=Genus,Length' 

used to give 

```
...
"data": [
        {
            "Genus": "Aaptosyax",
            "Length": 130.0
        },
        {
            "Genus": "Abactochromis",
            "Length": 11.5
        }
    ],
...
```

but now gives all fields back

> Is there a list of specific error messages that are useful that we should put in?

For now, would be nice to have helpful error message for:
- wrong argument class - e.g., string instead of integer for `limit` or `offset` 
- value too large: for `limit` & `offset`

**Comment by sckott** [comment]
any other thoughts @cboettig

**Comment by cboettig** [comment]
@EvilScott Looks like you've adjusted `docker.sh` for local testing, which is great but looks to me like it will no longer be suitable for actually deploying on a server.  For instance, it's not clear to me why your exposing the redis port; the sinatra container can access it through linking and having the redis port publicly exposed seems not ideal, right?  Also for deploying on a server we'd want to deploy on port `80` and not the randomly generated port from `-P`.  And of course we wouldn't then need the various testing bits that just attempt to open up the browser at the `/heartbeat` endpoint.  You're probably aware of all this but just wanted to note it here more for our reference.  

(Of course docker-compose would be suited for this, but when we started fig wasn't stable enough and had memory overrun issues).

**Comment by MollsReis** [comment]
@cboettig I have limited experience with Docker and I'm not personally familiar with your deployment strategy, so if there are any adjustments you need to make to the `Dockerfile` or `docker.sh` feel free to update them. :smile: 

I'm running on a Macbook Pro (as do most of the devs here) and docker does some weird acrobatics on Mac as far as ports and IPs via Vagrant and docker-machine which is why `docker.sh` opens `/heartbeat` in the manner it does.

**Comment by MollsReis** [comment]
@sckott I've been a bit under the weather lately so I've had some work on my main projects pile up (unfortunately, FBAPI isn't my top priority). I'll be addressing the issues you brought up very soon though.

**Comment by cboettig** [comment]
@EvilScott Right, I see what you're doing with the `docker-machine` call to get the ip and all and it makes perfect sense, I just think it's better to split out such platform-dependent parts of the script.  

And now sorry for a very stupid question, (@sckott 's probably already on top of this but it's escaping me): I'm not actually clear what you've done regarding the databases themselves.  Originally we separate mysql containers linking those to the fbapi container, as you saw (this potentially lets the mysql containers run on different nodes of a cluster, or float between networked nodes coreos/swarm style).  Now I'm having trouble following the logic here -- where does the API call the databases? Are you just running native instead of containerized copies of the databases?  

Related -- how should I deploy this to test (locally or otherwise?) Is it enough to just run the databases and then docker.sh script?

Also I see the `nginx` container has bitten the dust -- do you think that's unnecessary and we can serve directly from unicorn, or would you still imagine using a nginx reverse proxy? (Or maybe you just see that as belonging to the devops side and whoever is responsible for the server can handle that however the like -- we're not experienced enough to have clean separation between dev-ops and api development...  

(I'm completely on board with the comments you made re the logging and georeferencing stuff, where we were probably overengineering before we had a clear need.  And very nice stuff here, I could learn a lot from paying more attention to your code).

**Comment by MollsReis** [comment]
@cboettig I wasn't initially aware you were using Docker for a PROD deploy and I made a lot of changes up front because we weren't originally going to merge the fork back in. We can just bring back the old Docker files from git history if you guys want. For Unicorn, you should definitely use nginx for a reverse proxy for PROD deploys, though if you're running locally running `$ unicorn` (or just using `$ rackup config.ru`) is plenty. The `docker.sh` script actually checks for the existence of `docker-machine` (line 15) and uses localhost if it doesn't exist so `docker.sh` theoretically is platform independent. It is not intended to be a PROD deploy script however.

The API calls the DB via the models (see the two files in the `models` subdirectory. I implemented `activerecord` which is the industry standard ORM for Ruby. It abstracts away all raw SQL in favor of method calls, which also serves our purposes since we will be eventually migrating the FB DB to Postgres - now no queries need to be rewritten. As you can see from the changes, it also saves a lot of repetitious code.

The tests are now independent of the DB (we are testing the code, not the DB). Just `$ rake` to run the suite. It's just a starting point right now but its decent.

As for config, I followed a pretty common pattern to avoid committing sensitive credentials to the repo. You will find a `config_template.yml` file in the root. You will also notice `config.yml` is in the `.gitignore` file. The pattern is you copy and rename the config template (`$ cp config_template.yml config.yml`) then fill in the appropriate credentials for whatever environment you are pointing at. @sckott mentioned he got it running locally, you might ask him if he had to do anything special in addition.

**Comment by MollsReis** [comment]
Happy to answer you guys' questions, keep em coming! :smile_cat:

**Comment by MollsReis** [comment]
`http --verbose 'http://127.0.0.1:8080/species/?genus=Aborichthys&limit=2&fields=Genus,Length'` should give you the expected fields again with that latest commit

**Comment by MollsReis** [comment]
`/heartbeat` should now list model endpoints in correct casing (lowercase)

remember to `flushdb` in your `redis-cli` if you're running caching to see the latest updates

**Comment by MollsReis** [comment]
All non `GET` HTTP verbs will now 405. Was able to do this without using `sinatra/multi_route` (which wasn't handling wildcards correctly anyway).

**Comment by MollsReis** [comment]
@sckott for "value too large: for limit & offset" error, what is the allowed range for these fields?

**Comment by MollsReis** [comment]
404 and 500 error messages are now JSON per request

**Comment by sckott** [comment]
Limit 5000

**Comment by sckott** [comment]
No max offset

**Comment by MollsReis** [comment]
ok limit and offset are now required to be integer, and limit is max 5000

**Comment by MollsReis** [comment]
i think i've addressed all the concerns you had @sckott? if i need to make some adjustments to docker for @cboettig i can do that, but i think otherwise we are getting close to merge-ready

**Comment by sckott** [comment]
Awesome, nice work
- `http --verbose 'http://127.0.0.1:8080/species/?genus=Aborichthys&limit=2&fields=Genus,Length'` works nicely 
- heartbeat works great, thx
-  405's look good
- but some things not working now, e.g, `http 'http://127.0.0.1:8080/genera/8'` gives:
  
  ``````
  ```
  {
      "count": 0,
      "data": null,
      "error": {
          "message": "undefined method `<=' for nil:NilClass"
      },
      "returned": 0
  }
  ```
  ``````
  
   I think this affects all the routes that take pattern `/route/{ID}`, just `genera`/`species`/`faoarref`/`faoareas`

**Comment by sckott** [comment]
@cboettig To run locally, just made  a `config.yaml` file, start mysql, start redis, then `unicorn`

**Comment by cboettig** [comment]
@sckott thanks!  um, the `config.yaml` is for what?  Can you point me to an example config?

**Comment by MollsReis** [comment]
@cboettig there is a `config_template.yaml` in the root of the project that you should base your config off of

@sckott i will look into that issue tomorrow

**Comment by sckott** [comment]
I shared my config with @cboettig ...

@EvilScott thanks!

**Comment by MollsReis** [comment]
@sckott issue should be fixed

**Comment by sckott** [comment]
yep, works, thanks

**Comment by sckott** [comment]
@EvilScott What are you thinking on tests. They are pretty slim right now. If you're short on time, we can add more after merging

**Comment by MollsReis** [comment]
@sckott completely up to you guys if you want more or not. when i rewrote the suite my focus was on a suite that testing code instead of libraries/db and didn't have a lot of external requirements.

what aspects of the API do you feel need additional coverage?

**Comment by sckott** [comment]
@EvilScott I'll go ahead and merge. Thanks for all your work on this!

**Comment by MollsReis** [comment]
No problem! Feel free to reach out with further questions and I'll make PR's for further enhancements!

**Comment by MollsReis** [comment]
Thanks to both you guys, @sckott  and @cboettig for the speedy feedback! :smile_cat:

**Comment by sckott** [comment]
Sounds good

**Pull request merged**